src/include/wiredtiger.in

```diff
@@ -4647,6 +4664,179 @@ struct __wt_file_handle {
+#define	WT_SS_OPEN_CREATE	0x1u
+#define	WT_SS_OPEN_READONLY	0x2u
+/* AUTOMATIC FLAG VALUE GENERATION STOP */
+#endif
+
+/*!
+ * The interface implemented by applications to provide a shared system
+ * implementation. This documentation refers to "object" and "bucket"
+ * to mean a "file-like object" and a "container of objects", respectively.
+ *
+ * <b>Thread safety:</b> WiredTiger may invoke methods on the WT_SHARED_STORAGE
+ * interface from multiple threads concurrently. It is the responsibility of
+ * the implementation to protect any shared data.
+ *
+ * Applications register implementations with WiredTiger by calling
+ * WT_CONNECTION::add_shared_storage.  See @ref custom_shared_storages for more
+ * information.
+ *
+ * @snippet ex_shared_storage.c WT_SHARED_STORAGE register
+ */
+struct __wt_shared_storage {
+	/*!
+	 * Return a location handle from a location string.
+         * A location string may encode a bucket name, or the equivalent for this
+         * shared storage system, authorization information for that bucket,
+         * naming prefixes to be used for objects in that bucket, etc.
+	 *
+	 * @errors
+	 *
+	 * @param shared_storage the WT_SHARED_STORAGE
+	 * @param session the current WiredTiger session
+	 * @param location the location string
+	 * @param[out] location_handle the allocated handle
+	 */
+	int (*ss_location_handle)(WT_SHARED_STORAGE *shared_storage,
+	    WT_SESSION *session, const char *location, void **location_handle);
+
+	/*!
+	 * Free a location handle created by WT_SHARED_STORAGE::ss_location_handle.
+	 *
+	 * @errors
+	 *
+	 * @param shared_storage the WT_SHARED_STORAGE
+	 * @param session the current WiredTiger session
+	 * @param location_handle the handle to be freed.
+	 */
+	int (*ss_location_handle_free)(WT_SHARED_STORAGE *shared_storage,
+	    WT_SESSION *session, void *location_handle);
+
+	/*!
+	 * Return a list of object names for the given location.
+	 *
+	 * @errors
+	 *
+	 * @param shared_storage the WT_SHARED_STORAGE
+	 * @param session the current WiredTiger session
+	 * @param location_handle the location to list
+	 * @param[out] objlist the method returns an allocated array of
+	 *    individually allocated strings, one for each object in the location.
+	 * @param[out] countp the number of entries returned
+	 */
+	int (*ss_location_list)(WT_SHARED_STORAGE *shared_storage,
+	    WT_SESSION *session, void *location_handle,
+	    char ***objlist, uint32_t *countp);
+
+	/*!
+	 * Free memory allocated by WT_SHARED_STORAGE::location_list.
+	 *
+	 * @errors
+	 *
+	 * @param shared_storage the WT_SHARED_STORAGE
+	 * @param session the current WiredTiger session
+	 * @param objlist array returned by WT_SHARED_STORAGE::location_list
+	 * @param count count returned by WT_SHARED_STORAGE::location_list
+	 */
+	int (*ss_location_list_free)(WT_SHARED_STORAGE *shared_storage,
+	    WT_SESSION *session, char **objlist, uint32_t count);
+
+	/*!
+	 * Return if the named object exists in the location.
```

Comment:
On first reading, I interpreted this as "if the object exists, return" leaving me wondering what happens if it doesn't exist :-)

Reply:
Maybe add "whether or not" (and to similar file system calls if they're described the same way).